src/client/debugger/jupyter/kernelDebugAdapter.ts

```diff
@@ -0,0 +1,331 @@
+};
+
+const debugResponse = (message: DebugProtocol.Response): KernelMessage.IDebugReplyMsg => {
+    return {
+        channel: 'control',
+        header: {
+            msg_id: randomBytes(8).toString('hex'),
+            date: new Date().toISOString(),
+            version: '5.2',
+            msg_type: 'debug_reply',
+            username: 'vscode',
+            session: randomBytes(8).toString('hex')
+        },
+        metadata: {},
+        parent_header: {},
+        content: {
+            seq: message.seq,
+            type: 'response',
+            request_seq: message.request_seq,
+            success: message.success,
+            command: message.command,
+            message: message.message,
+            body: message.body
+        }
+    };
+};
+
+export class KernelDebugAdapter implements vscode.DebugAdapter {
+    private readonly fileToCell = new Map<string, vscode.NotebookCell>();
+    private readonly cellToFile = new Map<string, string>();
+    private readonly sendMessage = new vscode.EventEmitter<vscode.DebugProtocolMessage>();
+    private readonly messageListener = new Map<
+        number,
+        Kernel.IControlFuture<KernelMessage.IDebugRequestMsg, KernelMessage.IDebugReplyMsg>
+    >();
+
+    onDidSendMessage: vscode.Event<vscode.DebugProtocolMessage> = this.sendMessage.event;
+
+    constructor(
+        private session: vscode.DebugSession,
+        private notebookDocument: vscode.NotebookDocument,
+        private readonly jupyterSession: IJupyterSession,
+        private cellMap: IDebuggingCellMap
+    ) {
+        const iopubHandler = (msg: KernelMessage.IIOPubMessage) => {
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            if ((msg.content as any).event === 'stopped') {
+                console.error(msg);
+                this.sendMessage.fire(msg.content);
+            }
+        };
+        this.jupyterSession.onIOPubMessage(iopubHandler);
+
+        void this.dumpCellsThatRanBeforeDebuggingBegan();
+    }
+
+    async handleMessage(message: DebugProtocol.ProtocolMessage) {
+        // intercept 'setBreakpoints' request
+        if (message.type === 'request' && (message as DebugProtocol.Request).command === 'setBreakpoints') {
+            const args = (message as DebugProtocol.Request).arguments;
+            if (args.source && args.source.path && args.source.path.indexOf('vscode-notebook-cell:') === 0) {
+                await this.dumpCell(args.source.path);
+            }
+        }
+
+        // after attaching, send a 'debugInfo' request
+        // reset breakpoints and continue stopped threads if there are any
+        // we do this in case the kernel is stopped when we attach
+        // This might happen if VS Code or the extension host crashes
+        if (message.type === 'request' && (message as DebugProtocol.Request).command === 'attach') {
+            await this.debugInfo();
+        }
+
+        // after disconnecting, hide the breakpoint margin
```

Comment:
Can you elaborate a bit in this comment? I assume detaching is the same as hitting the 'debug' button on the toolbar. Can the user also cause this by just hitting the stop button during debugging? 

Seems weird to me. Stopping during debugging should just interrrupt the cell I think.

Reply:
normally nothing needs to happen here. But since vscode has a command that shows and hides the breakpoint margin for notebooks, we need to hide it at some point. And the best time to do it is on the disconnect command.